dist/server/PublishJob.js

```diff
@@ -0,0 +1,413 @@
+    })
+    .catch(err => {
+      console.error('Publication error, could not split file:', err.message);
+      this.error = true;
+      this.setStatus(err.message);
+    })
+
+    // download all assets
+    // check existing folder structure
+    .then(() => {
+      if(this.isStopped()) {
+        return;
+      }
+      return this.readOperations();
+      // const total = this.tree.actions.length;
+      // let downloaded = 0;
+      // this.setStatus(`Downloading ${total} files for ${file.name}`);
+      // console.log('tree:', this.tree);
+      // return Promise.all(
+      //   this.tree.actions.map(action => {
+      //     console.log('download start', action.displayName);
+      //     return this.unifile.readFile(this.session.unifile, action.src.service, action.src.path)
+      //     .then(buffer => {
+      //       console.log('download end', action.displayName);
+      //       downloaded++;
+      //       this.setStatus(`Downloading ${downloaded}/${total} files for ${file.name}`);
+      //       console.log('download ok', action.src);
+      //       action.buffer = buffer
+      //       return action;
+      //     })
+      //     .catch(err => {
+      //       console.warn('could not download file', action, err);
+      //       return Promise.resolve(action);
+      //     })
+      //   })
+      // )
+    })
+    .catch(err => {
+      // FIXME: will never go through here
+      console.error('Publication error, could not download files:', this.tree.actions.map(action => action.displayName).join(', '), '. Error:', err.message);
+      this.error = true;
+      this.setStatus(err.message);
+    })
+
+    // write and upload all files in a batch operation
+    .then(([statCss, statJs, statAssets, ...assets]) => {
+      if(this.isStopped()) {
+        return;
+      }
+      return this.writeOperations(statCss, statJs, statAssets, ...assets)
+    })
+    .catch(err => {
+      console.error('An error occured in unifile batch', err.message, err);
+      this.error = true;
+      this.setStatus(err.message);
+    })
+
+    // all operations done
+    .then(() => {
+      if(this.isStopped()) {
+        return;
+      }
+      this.setStatus(this.getSuccessMessage());
+      this.success = true;
+    })
+  }
+
+  readOperations() {
+    this.setStatus(`Creating folders: <ul><li>${this.cssFolder}</li><li>${this.jsFolder}</li><li>${this.assetsFolder}</li></ul>`);
+
+    // do not throw an error if the folder is not found, this is what we want to test
+    // instead catch the error and do nothing so that the result is null in .then(stat
+    const preventErr = promise => promise.catch(err => {
+      if(err.code != 'ENOENT') {
+        throw err;
+      }
+    });
+
+    // start by testing if the folders exist before creating them
+    // then download all assets
+    // FIXME: should use unifile's batch method to avoid conflicts or the "too many clients" error in FTP
+    //return Promise.all([
+    return sequential([
```

Comment:
As previously discussed you should do that. But maybe you want to clean that in another PR

Reply:
Yes as you told me it is better to test if a file exist and by writing and checking the error

But I would rather do that in an other PR because this current solution is the result of many problems with unifile batch and I am not comfortable changing it now